### ISSUE TITLE
Semantics of parameters in functions

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1594]
**Reported by fcasella on 28 Oct 2014 17:10 UTC**
The Modelica specification at section 3.8 defines the variability of expressions (constant, parameter, discrete-time, continuous-time). Sections 4.4.4 and 8.6 also discuss properties of variables with different variability. All these definitions apply to variables that eventually end up in the list of variables of a flattened model, though this is not stated explictly anywhere. It is not clear at a all if they also hold for variables inside a function.

As a consequence, it is not clear at all what is the semantics of parameters in functions.

Consider this example

```
package TestFunctionParameter
  function f
    input Real x;
    output Real y;
  protected
    parameter Real p = x * 2;
  algorithm
    y := p;
  end f;

  model M
    Real x = time;
    Real y = f(x);
  end M;
end TestFunctionParameter;
```

Both Dymola and OpenModelica accept this code and simulate model M, with the result being that y = time\*2.

My _a posteriori_ interpretation is that upon each call of function f, the parameter p is initialized to the value of its binding expression, and then it cannot be reassigned in the body of the function. In other words, it is constant _during the execution of the function body_, not _over time_, as Modelica functions are memoryless and do not have a notion of time. However, this is not stated anywhere in the specification.

I think the subject should be clarified. Incidentally, it is also possible to declare discrete variables in functions (the specification does not forbid it and tools accept it), but they probably make no sense at all, as there is no concept of time insided functions.

Summing up, I would suggest to write a special section about variability _inside functions_, e.g. in Section 4.4.4, with restrictions w.r.t. the variables in models, and to decide once and for all what meaning (if any) should parameters in functions have.

See also #1593

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1594

**Comment by modelica-trac-importer** [comment]
**Comment by sjoelund&#x2e;se on 28 Oct 2014 17:49 UTC**
Replying to [ticket:1594 fcasella]:
> Both Dymola and OpenModelica accept this code and simulate model M, with the result being that y = time\*2.

Seems to be a bug in OM. We do not allow using the inputs to bind parameter fields in records for example. It is intended to not allow this binding (probably because array dimensions can be input expressions inside functions and something got messed up).

> Incidentally, it is also possible to declare discrete variables in functions (the specification does not forbid it and tools accept it), but they probably make no sense at all, as there is no concept of time insided functions.

All variables are discrete-time inside functions (12.2).

**Comment by modelica-trac-importer** [comment]
**Comment by fcasella on 28 Oct 2014 18:29 UTC**
Replying to [comment:1 sjoelund.se]:
> All variables are discrete-time inside functions (12.2).

I read that sentence, but its meaning is unclear to me. There is no "time" inside functions, all variables come into existence when the function is called, and cease to exist upon returning to the caller. The definition of "discrete-time" variables that is given elsewhere in the spec makes explicit reference to the fact that they only change at events, which IMHO is totally confusing in this context.

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 28 Oct 2014 19:20 UTC**
MapleSim (correctly I believe) gives an error:

 Error, modification of parameter `f.p` must be a constant or parameter expression, but contains variable `f.x`

**Comment by modelica-trac-importer** [comment]
**Comment by msasena on 28 Oct 2014 19:27 UTC**
LMS Amesim gives a similar error about variability conflicts.

**Comment by modelica-trac-importer** [comment]
**Comment by hubertus on 28 Oct 2014 20:04 UTC**
Based on the variability argument an consistency with variability in an equation section this shoudl probably be an error. If someone can point to an unambigous rule in the specification clarifying this, this ticket can be closed. If not, a clarifying rule should be added.

**Comment by modelica-trac-importer** [comment]
**Comment by beutlich on 29 Oct 2014 10:59 UTC**
SimulationX 3.6.3 gives same result as Dymola and OM here.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 29 Oct 2014 12:21 UTC**
JModelica.org also accepts this, but I'd say that is a bug. Probably because of array sizes, like OpenModelica.

**Comment by modelica-trac-importer** [comment]
**Comment by jmattsson on 10 Dec 2014 15:37 UTC**
Not discussed at design meeting 84.

**Comment by modelica-trac-importer** [comment]
**Comment by dietmarw on 15 Apr 2015 14:33 UTC**
Ticket retargeted after milestone closed

**Comment by thorade** [comment]
Is this question only about parameters inside protected section?
Is the following a reasonable use case?

```modelica
package TestFunctionParameter
  function myFun
    input Real x;
    output Real y;
    parameter Real p = 1;
  algorithm 
    y := p*x;
  end myFun;

  model M
    function hereFun = myFun(p=3);
    Real x = time;
    Real y = hereFun(x);
  end M;
end TestFunctionParameter;
```

**Comment by HansOlsson** [comment]
Thus closing.
